src/pyflakes/checks.rs

```diff
@@ -31,6 +47,131 @@ pub fn string_dot_format_invalid(literal: &str, location: Range) -> Option<Check
     }
 }
 
+// F522
+pub fn string_dot_format_extra_named_arguments(
+    literal: &str,
+    keywords: &[Keyword],
+    location: Range,
+) -> Option<Check> {
+    if has_star_star_kwargs(keywords) {
+        return None;
+    }
+
+    let keywords = keywords.iter().filter_map(|k| {
+        let KeywordData { arg, .. } = &k.node;
+        arg.clone()
+    });
+
+    match FormatSummary::try_from(literal) {
```

Comment:
Maybe instead we have these all take `FormatSummary`, and in the checker, we have something like:

```rust
if let Ok(summary) = FormatSummary::try_from(literal) {
    if self.settings.enabled.contains(&CheckCode::F524) {
        if let Some(check) =
            pyflakes::checks::string_dot_format_missing_argument(
                value, args, keywords, location,
            )
        {
            self.add_check(check);
        }
    }
    ...
}
```

We could even gate that behind `self.settings.enabled.contains(&CheckCode::F524) | ...` to only compute it if at least one of those checks is enabled.

Reply:
Yeah, that sounds good. Gate the parsing/building and do it once.

The happy path is that `F521` and some of `F522-F525` are enabled together and it will be very rare that **only** F521 is enabled - so a little more performant in the default case to go with something like:

```rust
match FormatSummary::try_from(literal) {
    Err(e) => ..., // Add an F521 check if it's enabled
    Ok(summary) => {
        // Call the other enabled checks using the summary
    }
}
```